### PR TITLE
Allow validation based on halconfig context

### PR DESCRIPTION
**Pull request by lwander**
This is necessary for validations that occur across accounts, i.e. Kubernetes depends on Docker Registries.

@duftler @ttomsu @jtk54 PTAL

### CODE REVIEW

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * A reference to a class in halconfig.
+ *
+ * @param <T> Type of the reference
+ */
+@Data
+public class Reference<T> {
+  /**
+   * The optional value of the reference. May be null.
+   */
+  private T value;
```
Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java]: Why not use `Optional`?

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Updateable.java [hunk 2]
```diff
@@ -75,45 +71,45 @@
   /**
    * Validate a given field without updating it.
    *
-   * @param fieldName Name of the field to be validated.
-   * @param valueType The type of value being validated (can't be inferred when value is null).
+   * @param field The Field being valiated.
```
Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Updateable.java]: s/valiated/validated.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java (hunk 1 above)
Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java]: I believe that would remove the need for `valueType`.

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/FieldReference.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import lombok.Data;
+
+/**
+ * A reference to a field in halconfig.
+ * @param <T> The type of the update
+ */
+@Data
+public class FieldReference<T> extends Reference<T> {
+  public FieldReference() {}
```
Comment (ttomsu) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/FieldReference.java]: Alternatively, lombok's `@NoArgsConstructor`

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java (hunk 1 above)
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java]: It's a good point, but this value just gets passed transparently to one of halconfig fields. Right now I never have to check if it's null, so wrapping it in `Optional` means extra lines of code.

Comment (jtk54) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java]: `Optional` also doesn't get rid of the extra field either -_-.

Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/Reference.java]: Unfortunately :(

File: halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/model/v1/UpdateableSpec.groovy [hunk 4]
```diff
@@ -168,3 +180,23 @@ public class AccountValidator extends Validator<MyAccount> {
...
+  Stream<String> validate() {
+    if (context.halyardVersion == null) {
+      return Stream.of(UpdateableSpec.ERROR_MESSAGE)
+    } else {
+      return null
+    }
+  }
+
+  @Override
+  boolean skip() {
```
Comment (ttomsu) [inline halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/model/v1/UpdateableSpec.groovy]: Unrelated to this PR, seems like this should be `enabled`

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/FieldReference.java (hunk 3 above)
Comment (lwander) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/model/v1/FieldReference.java]: Ooh, very nice.

File: halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/model/v1/UpdateableSpec.groovy (hunk 4 above)
Comment (lwander) [inline halyard-config/src/test/groovy/com/netflix/spinnaker/halyard/model/v1/UpdateableSpec.groovy]: Another reason was that (at least for me) I think of "enable" as something determined by the user, external to what they are enabling/disabling. In this case, the validator itself determines if it should skip a validation, so it feels weird to have it "enable/disable" itself.

**Pull request merged**